[SPARK-24372][build] Add scripts to help with preparing releases.

vanzin opened the pull request:
The "do-release.sh" script asks questions about the RC being prepared,
trying to find out as much as possible automatically, and then executes
the existing scripts with proper arguments to prepare the release. This
script was used to prepare the 2.3.1 release candidates, so was tested
in that context.

The docker version runs that same script inside a docker image especially
crafted for building Spark releases. That image is based on the work
by Felix C. linked in the bug. At this point is has been only midly
tested.

I also added a template for the vote e-mail, with placeholders for
things that need to be replaced, although there is no automation around
that for the moment. It shouldn't be hard to hook up certain things like
version and tags to this, or to figure out certain things like the
repo URL from the output of the release scripts.

Commit:
[SPARK-24372][build] Add scripts to help with preparing releases.

The "do-release.sh" script asks questions about the RC being prepared,
trying to find out as much as possible automatically, and then executes
the existing scripts with proper arguments to prepare the release. This
script was used to prepare the 2.3.1 release candidates, so was tested
in that context.

The docker version runs that same script inside a docker image especially
crafted for building Spark releases. That image is based on the work
by Felix C. linked in the bug. At this point is has been only midly
tested.

I also added a template for the vote e-mail, with placeholders for
things that need to be replaced, although there is no automation around
that for the moment. It shouldn't be hard to hook up certain things like
version and tags to this, or to figure out certain things like the
repo URL from the output of the release scripts.

SparkQA commented:
**[Test build #91590 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91590/testReport)** for PR 21515 at commit [`44454b6`](https://github.com/apache/spark/commit/44454b69ff87a19dd6aa20e22e569e3d0811109b).

SparkQA commented:
**[Test build #91590 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91590/testReport)** for PR 21515 at commit [`44454b6`](https://github.com/apache/spark/commit/44454b69ff87a19dd6aa20e22e569e3d0811109b).
 * This patch **fails RAT tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/91590/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/3871/
Test PASSed.

Commit:
Add RAT exclude.

SparkQA commented:
**[Test build #91673 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91673/testReport)** for PR 21515 at commit [`1366f23`](https://github.com/apache/spark/commit/1366f236aafec5014fb794f59e1f23252dd52621).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/3909/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/19/
Test PASSed.

SparkQA commented:
**[Test build #91673 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91673/testReport)** for PR 21515 at commit [`1366f23`](https://github.com/apache/spark/commit/1366f236aafec5014fb794f59e1f23252dd52621).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/91673/
Test FAILed.

felixcheung commented on dev/create-release/vote.tmpl:
```diff
@@ -0,0 +1,64 @@
+Please vote on releasing the following candidate as Apache Spark version {version}.
+
+The vote is open until {deadline} and passes if a majority of at least 3 +1 PMC votes are cast.
```
nit: personally I find `3 PMC +1 votes` more clear

vanzin replied:
The rules are a little bit more complicated than that. Updated.

felixcheung commented on dev/.rat-excludes:
```diff
@@ -106,3 +106,4 @@ spark-warehouse
 structured-streaming/*
 kafka-source-initial-offset-version-2.1.0.bin
 kafka-source-initial-offset-future-version.bin
+vote.tmpl
```
even if rat doesn't check, isn't vote.tmpl packaged into the source release this way?

felixcheung replied:
for example, `.git` is removed from release here https://github.com/apache/spark/blob/master/dev/create-release/release-build.sh#L157

vanzin replied:
Are you saying this file should not be packaged in the source release? Not sure I see why that would be the case. There's a lot of stuff in `.rat-excludes` that is still packaged.

felixcheung commented on dev/create-release/spark-rm/Dockerfile:
```diff
@@ -0,0 +1,89 @@
+WORKDIR /opt/spark-rm/output
+
+ARG UID
+RUN useradd -m -s /bin/bash -p spark-rm -u $UID spark-rm
```
does that mean the do-release script is run as the user "spark-rm"?
I thought it's generally best practice to gpg sign as yourself?

vanzin replied:
The gpg signature is completely unrelated to the user running the process.

Also, Docker being the weird thing it is, the user name in the container doesn't really matter much. It's running with the same UID as the host user who built the image, which in this case is expected the be the person preparing the release, and this will only really matter for files that are written back to the host's file system.

Commit:
Update vote count wording.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/68/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/3957/
Test PASSed.

SparkQA commented:
**[Test build #91734 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91734/testReport)** for PR 21515 at commit [`04f6371`](https://github.com/apache/spark/commit/04f6371a3aa0f03ba1c37b5b450bea69923388e9).

tgravescs commented:
I guess we will need to update the release process website once these go in?

Its not clear to me how the docker stuff applies here, are we recommending to create the release inside docker or leaving it as optional?  did you use it for 2.3.1 release?

vanzin commented:
We could update the docs once these get in.

I did not use docker for 2.3.1 but I plan to give it a shot for 2.1.3. The goal of the docker path is to always build Spark in a known environment, so things like R packages and everything else are the same. It could also help folks who use other OSes (although Sameer used MacOS for 2.3.0 and it seemed to work fine, aside from some warnings from gnu tar when decompressing the tarballs on Linux).

SparkQA commented:
**[Test build #91734 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91734/testReport)** for PR 21515 at commit [`04f6371`](https://github.com/apache/spark/commit/04f6371a3aa0f03ba1c37b5b450bea69923388e9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/91734/
Test PASSed.

vanzin commented:
FYI I'm playing with building 2.1.3 using the docker-based approach, and there are a few things that I need to adjust before that works. I hope to update this PR later today.

Commit:
Changes mostly targeted at using the script for 2.1 builds.

- Allow mounting a local JDK (2.1 wants 1.7).
- Tweak profiles and builds to 2.1
- Restore support for Scala 2.10
- set up JAVA_HOME in the docker container

I also expanded "dry run" mode a bit to build one of the binary
packages, and also the documentation. The build documentation is
now also kept in the output directory along with the other
artifacts.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/155/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/4046/
Test PASSed.

SparkQA commented:
**[Test build #91869 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91869/testReport)** for PR 21515 at commit [`48eda71`](https://github.com/apache/spark/commit/48eda71029bb414f404312fbdecf7a35d883660c).

SparkQA commented:
**[Test build #91869 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91869/testReport)** for PR 21515 at commit [`48eda71`](https://github.com/apache/spark/commit/48eda71029bb414f404312fbdecf7a35d883660c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/91869/
Test PASSed.

tgravescs commented on dev/create-release/do-release.sh:
```diff
@@ -0,0 +1,59 @@
+SELF=$(cd $(dirname $0) && pwd)
+. "$SELF/release-util.sh"
+
+while getopts "bn" opt; do
```
it would be nice to have a high level description in the script just saying this does a release which does things like tag, build, etc and pushes things to the asf spark repo.

tgravescs commented on dev/create-release/release-util.sh:
```diff
@@ -0,0 +1,219 @@
+  # Check if the RC already exists, and if re-creating the RC, skip tag creation.
+  RELEASE_TAG="v${RELEASE_VERSION}-rc${RC_COUNT}"
+  SKIP_TAG=0
+  if check_for_tag "RELEASE_TAG"; then
```
this needs to be $RELEASE_TAG

tgravescs commented:
not sure if this was something in my env but somehow it pushed next version as 2.2-3-snapshot.
https://github.com/apache/spark/commits/branch-2.2
I manually fixed it, and am building now but skipped the push of tag since it was already there. The build had failed due to JAVA_HOME not set. 
If I reproduce I'll debug.

vanzin commented:
> pushed next version as 2.2-3-snapshot

Hmm must be something I introduced in the last changes (2.3.2 seems correct). Sorry about that.

> failed due to JAVA_HOME not set

I think that might happen if you don't use the `-j` option I added, but let me double check.

Commit:
Small fixes.

vanzin commented:
I was able to run without `-j`, no problems with JAVA_HOME...

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/4178/
Test PASSed.

SparkQA commented:
**[Test build #92036 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92036/testReport)** for PR 21515 at commit [`7fe133a`](https://github.com/apache/spark/commit/7fe133afac729e8ebc2bba864ead03345ea0ed7c).

Commit:
Another version number fix.

SparkQA commented:
**[Test build #92044 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92044/testReport)** for PR 21515 at commit [`faab8ae`](https://github.com/apache/spark/commit/faab8ae167c06b61340b2d3aa9cfb1ce71435d8c).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/289/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/4183/
Test PASSed.

Commit:
Un-hardcode dry run mode with docker.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/4184/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/290/
Test PASSed.

SparkQA commented:
**[Test build #92045 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92045/testReport)** for PR 21515 at commit [`82ae00e`](https://github.com/apache/spark/commit/82ae00e61dfbe40e75dbcd3d224fcd5c3f5f74a7).

SparkQA commented:
**[Test build #92036 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92036/testReport)** for PR 21515 at commit [`7fe133a`](https://github.com/apache/spark/commit/7fe133afac729e8ebc2bba864ead03345ea0ed7c).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/92036/
Test FAILed.

Commit:
Some updates for Java 7.

Seems like Maven Central updated something over the weekend, and Java 7
was having trouble connecting; needed some extra SSL-related options.

I also made the build sequential, which makes it slower, but avoids some
issues with parallel maven processes that I was hitting when building
2.1.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/296/
Test PASSed.

SparkQA commented:
**[Test build #92051 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92051/testReport)** for PR 21515 at commit [`5771354`](https://github.com/apache/spark/commit/57713541abd926d5697da7ef90f843eaaf9ce67f).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/4190/
Test PASSed.

SparkQA commented:
**[Test build #92044 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92044/testReport)** for PR 21515 at commit [`faab8ae`](https://github.com/apache/spark/commit/faab8ae167c06b61340b2d3aa9cfb1ce71435d8c).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/92044/
Test FAILed.

SparkQA commented:
**[Test build #92045 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92045/testReport)** for PR 21515 at commit [`82ae00e`](https://github.com/apache/spark/commit/82ae00e61dfbe40e75dbcd3d224fcd5c3f5f74a7).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/92045/
Test PASSed.

Commit:
More fixes.

- Override SBT_OPTS for Java 7
- Allow executins individual steps in docker
- Fix docker image so that changing what to install also forces
  an "apt update" to happen.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/4197/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/303/
Test PASSed.

SparkQA commented:
**[Test build #92058 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92058/testReport)** for PR 21515 at commit [`a39933e`](https://github.com/apache/spark/commit/a39933eef41370c07cf4ac70665943dfe0ae00a9).

Commit:
Even more fixes.

- Some bugs I introduced in last commit.
- Fix release-tag.sh with Java 1.7.
- Expand dry run to tag and publish steps.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/306/
Test PASSed.

SparkQA commented:
**[Test build #92061 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92061/testReport)** for PR 21515 at commit [`b6d4c70`](https://github.com/apache/spark/commit/b6d4c705b7aea7b55a1618892cba12bedb1cd6c5).

SparkQA commented:
**[Test build #92051 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92051/testReport)** for PR 21515 at commit [`5771354`](https://github.com/apache/spark/commit/57713541abd926d5697da7ef90f843eaaf9ce67f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/92051/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/92058/
Test PASSed.

SparkQA commented:
**[Test build #92061 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92061/testReport)** for PR 21515 at commit [`b6d4c70`](https://github.com/apache/spark/commit/b6d4c705b7aea7b55a1618892cba12bedb1cd6c5).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/92061/
Test FAILed.

Commit:
Avoid silent failure if user.name is not set.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/326/
Test PASSed.

SparkQA commented:
**[Test build #92097 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92097/testReport)** for PR 21515 at commit [`745342d`](https://github.com/apache/spark/commit/745342d499ef290d5ce5484bcb2349d2be57a2fb).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/4220/
Test PASSed.

tgravescs commented:
just a note I successfully used these scripts with docker with dry-run option with spark 2.2.2 rc1.

tgravescs commented:
at this point should we just merge these and make add anything else that comes up in other prs ?

vanzin commented:
SGTM.

tgravescs commented:
ok, +1, will merge shortly

tgravescs commented:
merged, did you have a jira to update the documents to match?

tgravescs commented:
Also thanks again for doing these scripts, the docker was much easier then me trying to get all dependencies on rhel!

vanzin commented:
I did not have a task filed for the doc updates.

cloud-fan commented on dev/create-release/do-release-docker.sh:
```diff
@@ -0,0 +1,143 @@
+  fi
+done
+
+GPG_KEY_FILE="$WORKDIR/gpg.key"
```
how shall we generate this `gpg.key`?

cloud-fan commented:
I hit an issue when using this script to prepare 2.4.0 rc1
```
useradd: UID 0 is not unique
The command '/bin/sh -c useradd -m -s /bin/bash -p spark-rm -u $UID spark-rm' returned a non-zero code: 4
```

Any ideas about how to resolve it?

felixcheung commented:
UID already exists?

cloud-fan commented on dev/create-release/do-release-docker.sh:
```diff
@@ -0,0 +1,143 @@
+$GPG --export-secret-key --armor "$GPG_KEY" > "$GPG_KEY_FILE"
+
+run_silent "Building spark-rm image with tag $IMGTAG..." "docker-build.log" \
+  docker build -t "spark-rm:$IMGTAG" --build-arg UID=$UID "$SELF/spark-rm"
```
So we need to do `export UID=xxx` before running this script?

cloud-fan replied:
got it. This is a system variable. So we can't run this script with root user...

cloud-fan commented:
Problem solved. This script assumes the current user can run docker command without `sudo`, while it may not be true. But we can't run this script with root user as we can't add a root user in the docker container.

Following the steps in https://docs.docker.com/install/linux/linux-postinstall/#manage-docker-as-a-non-root-user solves it.
